### DISCUSSION TITLE
Let's break backwards compatibility!

### DISCUSSION THREAD

**Discussion by SheepTester** [nbuilding/N-lang#258] [Ideas]
N shall crawl into a cave (releasing only 0.x versions) until it is more finalised and ready, then it shall emerge as N 2.0; how exciting!

Discuss some ideas for improving N here!

**Comment by Ashvin-Ranjan** [comment]
Here are some things to discuss!
- [ ] Syntax Improvements
- [ ] Function Improvements
- [ ] Language Library Mappings
- [ ] Preformance Improvoments
- [ ] Built-in types
- [ ] Mutability
- [ ] Syntax Consistency

Let me know if you want more added!

**Reply by SheepTester** [reply]
What syntax improvements

**Reply by Ashvin-Ranjan** [reply]
https://github.com/nbuilding/N-lang/discussions/258#discussioncomment-1178517

**Comment by Ashvin-Ranjan** [comment]
Syntax Improvements Thread:

**Reply by ky28059** [reply]
Hmm, I'll just paste some things I've said here:
![image](https://user-images.githubusercontent.com/60120929/129435560-245b3537-073c-442c-8a0f-5018fd88611f.png)
And I still think that N needs a more explicit `void` type than empty tuple.

**Reply by Ashvin-Ranjan** [reply]
Do you think we should have a function keyword or just stick with anonymous functions

**Reply by SheepTester** [reply]
Well it's pretty common in JS to define functions using function expressions

```js
export const Sidebar = ({ collapsed }) => {
  return (
    <div class={collapsed ? 'collapsed' : ''}>
      {/* ... */}
    </div>
  )
}
```

**Reply by SheepTester** [reply]
> And I still think that N needs a more explicit `void` type than empty tuple.

Well since N functions are pure, there aren't any void functions, but there are functions that return `cmd[()]`. Would you rather it be more verbose, kind of like TypeScript `Promise<void>`? Alternatively, functions with no `return` statements could have an implicit return type of `cmd[()]`, so you wouldn't have to write it at all (like doing `async function main () {` in TS)

**Reply by Ashvin-Ranjan** [reply]
I prefer the anonymous function way but if you want to then sure

**Reply by SheepTester** [reply]
I think we could maybe make function expressions easier to write. They can just be a series of `pattern: type ->` followed by a return type (which can be implied from `return` statements, the context the function is being used in, and/or using the implicit `cmd[()]` as above.

The function body can either be curly brackets enclosing a series of statements or a colon followed by an expression (syntax for this expression shorthand isn't final).

`pattern` is any destructuring pattern, so it can just be the argument name, like

```ts
// Return type is implied because n * n is a float
let square = n: float -> : n * n

// More verbose equivalent
let square: float -> float = (n: float) -> float {
  return n * n
}
```

But it can also be a destructuring pattern like `(a, b)` to destructure a tuple. Normally, you would have to write `(a, b): (int, int)` when destructuring a tuple whilst also giving the type annotations, which is a bit annoying because the types aren't right next to the argument names. Instead, we could also allow you to add types *within* a pattern:

```ts
let add = (a: int, b: int) -> : a + b

// More verbose equivalent
let add: (int, int) -> int = ((a, b): (int, int)) -> int {
  return a + b
}
```

**Reply by SheepTester** [reply]
> I prefer the anonymous function way but if you want to then sure

Who wants what? I prefer function expressions as well; it makes it more clear that functions are also data types

**Reply by Ashvin-Ranjan** [reply]
Also, I think that the enum destructuring / enum creation syntax really needs to be fixed because we have c-style function calls to make an enum but the destructuring name uses the only function call syntax. also, you don't do `<json.object thing>` you do `<object thing>`, which can cause problems.

**Reply by SheepTester** [reply]
Well `json.object` is just the constructor function; you can do

```ts
let whatever = json.object

if let <object weee> = whatever(...) {
```

And for user-defined types you can import them like

```ts
let { object } = imp "./json.n"
```

I think it's less verbose to be able to do `<object ...>` instead of `<json.object ...>` and it also works for non-native modules

**Reply by Ashvin-Ranjan** [reply]
The main problem with doing `<object ...>` instead of `<json.object ...>` can be a problem when two imported files have a similarly named enum, this also goes for enum constant names too.

**Reply by SheepTester** [reply]
It's not a problem because the type is known at type checking time

**Reply by Ashvin-Ranjan** [reply]
But it is syntactically ambiguous and it is slightly confusing. Also, how would this work?

```js
let a = imp "./test.n" // Contains an enum value called `test`
let b = imp "./test1.n" // Contains an enum value called `test`
print(test) // How does this work
```

**Reply by SheepTester** [reply]
The `object` in `if let <object ...> =` is not an expression; you can't just put any value in there, it has to be an identifier.

In your example, `test` is not defined

It's like how `.a` is both used to access a field from `{ a: str; c: int }` but also `{ b: bool; a: int }`.

**Comment by Ashvin-Ranjan** [comment]
Typescript features to adopt maybe

**Reply by Ashvin-Ranjan** [reply]
Ideas:
- Union Types
- Nullish access
- Nullish type casting
- `maybe` typed record fields being optional
- Better type constricting (If a `maybe` is evaluated to not `none` then cast it to its regular type)
- Interfaces
- Type mappings from java modules to N

**Reply by SheepTester** [reply]
> * Union Types
> * Nullish access
> * Nullish type casting

We already have these

**Reply by Ashvin-Ranjan** [reply]
Nullish access does not exist for records or classes
Where did we implement Nullish type casting and Union Types

**Reply by SheepTester** [reply]
The `yes` function casts a type to the nullish type (`maybe`)

You can use custom types (enums) as union types

Ah, I was thinking of the `a[b]` operator

**Comment by Ashvin-Ranjan** [comment]
Fixing `json` because it is really hard to use `json`s in N rn

**Reply by SheepTester** [reply]
I was working on a JSON parsing library that would've made it very nice to parse and create JSON

**Reply by Ashvin-Ranjan** [reply]
Can you finish it because `v1.3.0` came out and we can use it as a basis for `v2`

**Reply by SheepTester** [reply]
I believe it required comprehensive `match` expressions like

```rs
match (result) {
  <ok value> => ...,
  <err error> => ...,
}
```
or whatever the syntax is

I'm just waiting on the JS branch to finish first
